core/src/main/scala/org/apache/spark/memory/MemoryManager.scala

```diff
@@ -155,4 +164,217 @@ private[spark] abstract class MemoryManager extends Logging {
+      val maxToGrant =
+        math.min(numBytes, math.max(0, (maxExecutionMemory / numActiveTasks) - curMem))
+      // Only give it as much memory as is free, which might be none if it reached 1 / numTasks
+      val toGrant = math.min(maxToGrant, freeMemory)
+
+      if (curMem < maxExecutionMemory / (2 * numActiveTasks)) {
+        // We want to let each task get at least 1 / (2 * numActiveTasks) before blocking;
+        // if we can't give it this much now, wait for other tasks to free up memory
+        // (this happens if older tasks allocated lots of memory before N grew)
+        if (
+          freeMemory >= math.min(maxToGrant, maxExecutionMemory / (2 * numActiveTasks) - curMem)) {
+          return acquire(toGrant, taskAttemptId)
+        } else {
+          logInfo(
+            s"TID $taskAttemptId waiting for at least 1/2N of shuffle memory pool to be free")
+          wait()
+        }
+      } else {
+        return acquire(toGrant, taskAttemptId)
+      }
+    }
+    0L  // Never reached
+  }
+
+  /**
+   * Acquire N bytes of execution memory from the memory manager for the current task.
+   * @return number of bytes actually acquired (<= N).
+   */
+  private def acquire(numBytes: Long, taskAttemptId: Long): Long = synchronized {
+    val evictedBlocks = new ArrayBuffer[(BlockId, BlockStatus)]
+    val acquired = acquireExecutionMemory(numBytes, evictedBlocks)
+    // Register evicted blocks, if any, with the active task metrics
+    // TODO: just do this in `acquireExecutionMemory` (SPARK-10985)
+    Option(TaskContext.get()).foreach { tc =>
+      val metrics = tc.taskMetrics()
+      val lastUpdatedBlocks = metrics.updatedBlocks.getOrElse(Seq[(BlockId, BlockStatus)]())
+      metrics.updatedBlocks = Some(lastUpdatedBlocks ++ evictedBlocks.toSeq)
+    }
+    taskMemory(taskAttemptId) += acquired
+    acquired
+  }
+
+  /** Release numBytes bytes for the current task. */
+  def release(numBytes: Long, taskAttemptId: Long): Unit = synchronized {
+    val curMem = taskMemory.getOrElse(taskAttemptId, 0L)
+    if (curMem < numBytes) {
+      throw new SparkException(
+        s"Internal error: release called on $numBytes bytes but task only has $curMem")
+    }
+    if (taskMemory.contains(taskAttemptId)) {
+      taskMemory(taskAttemptId) -= numBytes
+      releaseExecutionMemory(numBytes)
+    }
+    notifyAll() // Notify waiters in tryToAcquire that memory has been freed
+  }
+
+  /** Release all memory for the current task and mark it as inactive (e.g. when a task ends). */
+  private[memory] def releaseMemoryForTask(taskAttemptId: Long): Unit = synchronized {
+    taskMemory.remove(taskAttemptId).foreach { numBytes =>
+      releaseExecutionMemory(numBytes)
+    }
+    notifyAll() // Notify waiters in tryToAcquire that memory has been freed
+  }
+
+  /** Returns the memory consumption, in bytes, for the current task */
+  private[memory] def getMemoryConsumptionForTask(taskAttemptId: Long): Long = synchronized {
```

Comment:
These all need to be renamed to refer to execution memory, since this class `MemoryManager` handles both execution and storage so it's not sure what kind of memory we're talking about. We actually also need to update a lot of comments too.